### PR TITLE
Minor performance optiomizations to improve sleep speed

### PR DESCRIPTION
**Pull request by Farseer2**
#### Summary
SUMMARY: Performance "Minor optimizations"

#### Purpose of change

Slightly improves some of the hotspots encountered when profiling #44157

#### Describe the solution
This commit includes 4 changes:
* Add an `unsafe` variant to get_submap_at, as there are _a lot_ of places in map.cpp which bound check multiple times. I converted most I found, but I'm sure there are still a few more.
* Move several tiny yet hot functions to headers so they will be inlined
* Use ref instead of copy in map::shoot
* Small std container usage changes (reserve upfront, count+insert -> insert)

#### Testing
Loaded up the given save, walked around a bit and slept
#### Additional notes
Not sure where to put this, but I think most of the class location should be made non-virtual.
All uses of pos(x/y/z) of it are the same, which begs the question why waste an indirect call whenever using any of them?
I can add this to the PR, but didn't know if this was done intentionally, so I haven't made this change yet.

### CODE REVIEW

File: src/mtype.cpp [hunk 1]
```diff
@@ -242,8 +242,8 @@ std::string mtype::get_description() const
 
 std::string mtype::get_footsteps() const
 {
-    for( const species_id &s : species ) {
-        return s.obj().get_footsteps();
+    if( species.size() ) {
```
Comment (Qrox) [inline src/mtype.cpp:245]:
```suggestion
    if( !species.empty() ) {
```

**Review by kevingranade — CHANGES_REQUESTED** [review]
I need to see some specific results showing that
``Move several tiny yet hot functions to headers so they will be inlined``
is significant.

The change to map::move_cost() is breaking in some tests and I strongly suspect it will trigger that debug mesage in-game as well.  Overall I'm very leary of sprinkling around a lot of calls that skip bounds checks, these changes will require some very careful review.

**Comment by Farseer2** [comment]
As for inlining, here's monster::pos (Which still won't be inlined as its virtual, pointing back to my question), which is a mov instruction when inlined:
image
The second one I inlined was is_hallucination, which should only be a single mov instruction (If inlined):
image

**Comment by jbytheway** [comment]
In some cases we can get virtual functions inlined by marking them (or the classes containing them) `final`.  That's a low-risk change which might alleviate your concern.

**Comment by Farseer2** [comment]
Well, if you're already making all the uses of location both final and the same, why not just de-virtualize it?

**Comment by mlangsdorf** [comment]
Can we also get some overall performance numbers?  How long does it take to sleep for 8 hours without these changes versus with them?

**Review by mlangsdorf — COMMENTED** [review]
These look like sensible changes to me but reviewing this is outside my comfort zone.

**Comment by kevingranade** [comment]
Okiedokie, I finally had a chance to review this line-by-line, and yes every instance of get_submap_at_unsafe() comes after an explicit inbounds() check.  That's a very clear "do less work" change, so I don't need to see explicit profiling, though I'll try to grab some anyway.

I think it will be worthwhile to go back and evaluate whether these callsites can instead just try to retrieve the submap and treat failures as a bounds check failure, but for now this is worth merging as is.

Thanks for the supporting information WRT the inlining, that's convincing to me.

**Comment by kevingranade** [comment]
Profiling shows a notable improvement from the map accesses.
Before:

After:


Specifically monster::move() and map::process_items() both picked up a few percent.

**Pull request merged**